pennylane/devices/execution_config.py

```diff
@@ -30,24 +32,34 @@
 ]
 
 
-@dataclass
+@dataclass(frozen=True)
 class ExecutionConfig:
     """
     A class to configure the execution of a quantum circuit on a device.
 
     See the Attributes section to learn more about the various configurable options.
     """
 
-    shots: Optional[Union[int, Tuple[int]]] = None
-    """The number of shots for an execution"""
+    # pylint: disable=too-many-instance-attributes
+
+    shots: int = None
```

Comment:
@albi3ro @timmysilv 
Have you already decided on how to deal with batches of circuits with different shots per circuit? 

Just thinking out loud... would it make more sense to put shots inside the `QNode`?

Reply:
Since `shots` is a 'per circuit' attribute (rather than a global execution option), this ties into the question of 'how can we provide execution config per tape'. This will be especially important for cases where we are batch executing multiple circuits, and want different shots per circuit.

Reply:
This is a good point. I think that some deeper discussion is very important at this stage.

If we want to batch execute `QuantumScript`s with the same `ExecutionConfig`(EC) for all `QuantumScript`s and also keep shots within `ExecutionConfig`, we could check the shape of `shots` that are passed in when the EC is created and initialize accordingly. However, this significantly convolutes `MeasurementProcess.shape` as it relies on EC to provide the shots and assumes that the given shots are for a single execution. This can be changed as well, but that should probably be a different PR

Reply:
The way I see it, a `tape` is **WHAT** we want to execute (gates, pulses, measurements...) and `ExecutionConfig` should contain the **HOW** we want to execute it (gradient method, interface...).

Now the question is, is `shots` inside the WHAT or the HOW? In my opinion it is the former.

Reply:
I agree that it makes more sense to store shot information on the `QuantumScript` rather than `ExecutionConfig`. I'd like to get everyone's opinion as well before I potentially start making changes.